internal/core/metadata/rest_deviceprofile.go

```diff
@@ -355,6 +367,16 @@ func restAddProfileByYamlRaw(
 		return
 	}
 
+	err = validateDeviceProfileStruct(dp)
```

Comment:
I don't see any validation of JSON Profile. Should add there also?
After unmarshal, the code should be the same whether it came in via JSON or YAML.
Same for Update.

Reply:
Validation in JSON API will be triggered during unmarshall automatically
https://github.com/edgexfoundry/go-mod-core-contracts/blob/0746ea1719ff74015da8ef926f06a40e1e317cc2/models/deviceprofile.go#L75

Reply:
we only miss it in YAML API

Reply:
The current issue is that if the user create an invalid Profile via YAML API, it will always cause validation error during query.  If there is one invalid profile in database, GetAll profile will also always return validation error.

Reply:
Got it. Thanks!

Reply:
> Validation in JSON API will be triggered during unmarshall automatically
> https://github.com/edgexfoundry/go-mod-core-contracts/blob/0746ea1719ff74015da8ef926f06a40e1e317cc2/models/deviceprofile.go#L75

I tested this because I wanted to check your statement about our `Validator` interface not properly handling slices. It turns out this is true, which means this automatic validation during un-marshaling is not really doing all that much as both `DeviceCommands` and `CoreCommands` in `DeviceProfile` are both slices.

So this means with the addition of this PR, we're doing much stricter validation for YAML imports than for JSON imports...

Reply:
Do I need to modify this PR or open another PR to fix Validator for json imports?

Reply:
>  It turns out this is true, which means this automatic validation during un-marshaling is not really doing all that much as both DeviceCommands and CoreCommands in DeviceProfile are both slices.

No, when adding the JSON format DeviceProfile, the API invoke the **UnmarshalJSON** func, so it can validate both DeviceCommands and CoreCommands.
For example, resourceoperation's UnmarshalJSON func invoke the Validate function.
```
ro.isValidated, err = ro.Validate()
``` 
https://github.com/edgexfoundry/go-mod-core-contracts/blob/master/models/resourceoperation.go#L124 

> So this means with the addition of this PR, we're doing much stricter validation for YAML imports than for JSON imports...

It means we do some approach in this PR.

Reply:
@tonyespy In summary, we actually do the same level of validation in both JSON and YAML now.